### ISSUE TITLE
haproxy_stats.sh

### ISSUE THREAD

**Issue by jhan-aa** [anapsix/zabbix-haproxy#16]
I tested the haproxy_discover.sh and haproxy_stats.sh scripts on one of my agent server by following up the README.md file and everythin looks fine,
1. But from the zabbix latest data,  i always get "ERROR: haproxy socket is not writable"
   When i comment out below:
   # 
   # if [[ $QUERYING_METHOD == "SOCKET" && ! -w $HAPROXY_SOCKET ]]
   # then
   # echo "ERROR: haproxy socket is not writable"
   # exit 126
   # fi
2. The zabbix return me "ERROR: bad $pxname/$svname"
3. And i checked the CACHE_STATS_FILEPATH file, there is no data in the cache file.

Can you please help on that?

**Comment by noamik** [comment]
Are you sure that you enabled the stats socket in your haproxy configuration in the first place?

You should have a line in global similar to this one:
`stats socket /var/run/haproxy.sock mode 666 level user`

You can check whether your stats socket is working by typing:
`echo "show stat" | socat /var/run/haproxy.sock stdio`

You should get a lot of output in return. See this query of mine (it has LOTS of more lines):
`# pxname,svname,qcur,qmax,scur,smax,slim,stot,bin,bout,dreq,dresp,ereq,econ,eresp,wretr,wredis,status,weight,act,bck,chkfail,chkdown,lastchg,downtime,qlimit,pid,iid,sid,throttle,lbtot,tracked,type,rate,rate_lim,rate_max,check_status,check_code,check_duration,hrsp_1xx,hrsp_2xx,hrsp_3xx,hrsp_4xx,hrsp_5xx,hrsp_other,hanafail,req_rate,req_rate_max,req_tot,cli_abrt,srv_abrt,comp_in,comp_out,comp_byp,comp_rsp,lastsess,last_chk,last_agt,qtime,ctime,rtime,ttime
http_frontend,FRONTEND,,,0,5,16000,88316,17699358,147074969,0,0,389,,,,,OPEN,,,,,,,,,1,2,0,,,,0,0,0,11,,,,0,42377,3172,42767,0,0,,0,11,88316,,,0,0,0,0,,,,,,,,
https_frontend_inactive,FRONTEND,,,0,2,16000,109075,16911104,71933091,0,0,0,,,,,OPEN,,,,,,,,,1,3,0,,,,0,0,0,2,,,,0,100526,0,7590,959,0,,0,2,109075,,,0,0,0,0,,,,,,,,`

**Comment by jhan-aa** [comment]
Hi Thanks for your quick update.

Yes, the permission of the haproxy socket file is correct.

I can also get the correct output when i execute the echo "show stat" | socat $SOCKETPATH stdio manually

And my question is coming, before i execute above command manually, the SOCKET Cache file is bull, so zabbix can't get anydata, then it returns me "ERROR: bad $pxname/$svname" which can be approve from your code

---

get() {
  # $1: pxname/svname
  local _res="$(grep $1 ${CACHE_STATS_FILEPATH})"
  if [ -z "${_res}" ]
  then
    echo "ERROR: bad $pxname/$svname"
    exit 127
  fi

---

So the problem is the socket cache file can not be updated continuslly at my end. But when i manually run the command "show stat" | socat ......   The socket file  will be filled one time. please let me know how can i fix this, did you encounter at your end?

**Comment by jhan-aa** [comment]
Sorry i gorget i thing, i socket permission is 755, might be the cause? should i change it to 766?

**Comment by noamik** [comment]
Well, the error you got was that the script couldn't write the socket. If you can do it on the command line, you will need to find the difference in permissions between script execution and your command line execution. The most obvious issue I can think of would be you trying it as root while the script is executed by zabbix as zabbix user, but there could as well be other factors at play, like selinux. You will need to debug this by yourself. Uncommenting the code as you did won't help, since the line is there for a reason.

**Comment by jhan-aa** [comment]
Hi Really thanks for your update, my problem was resolved.

**Comment by noamik** [comment]
Feel free to share the solution if you think it's something other users could stumble into as well ...

**Comment by darkshade9** [comment]
I stumbled upon this problem and would like a solution if possible.

Socket is read/writable to everyone (666 permissions)
srw-rw-rw-. 1 root   root   0 Aug 12 13:39 haproxy.sock

ERROR: haproxy socket is not writable

Edit: Fixed it myself.  Ensure your haproxy socket path is defined in both haproxy_discovery and haproxy_stats are the same.

**Comment by anapsix** [comment]
@darkshade9 are you passing your custom socket path to the script?

**Comment by darkshade9** [comment]
@anapsix Yes, I'm using it in a Docker containerized environment and I'm using /tmp/haproxy.sock as the path

**Comment by anapsix** [comment]
@darkshade9: custom socket path can be passed to both scripts as a first argument; and so there is nothing to define, as long as you customize it in Zabbix Web UI via host/template variable or, if you are playing with cli, pass it as first argument.

**Comment by AlexeyVshk** [comment]
Hi all,
How have you solved the problem? I've got the same result --> ERROR: haproxy socket is not writable.
1. The permission of the haproxy socket file is correct.
2. I can also get the correct output when i execute the echo "show stat" | socat /var/lib/haproxy/stats stdio !
3. I've corrected the path inside the scripts --> 
- HAPROXY_SOCK in the haproxy_discovery.sh ;
- HAPROXY_SOCKET in the haproxy_stats.sh ;

????
